src/EditorFeatures/Core/IntelliSense/NavigationActionFactory.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Host;
+using Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.QuickInfo;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.GoToDefinition;
+using Microsoft.CodeAnalysis.Navigation;
+using Microsoft.CodeAnalysis.QuickInfo.Presentation;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Threading;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense;
+
+internal sealed class NavigationActionFactory(
+    Document document,
+    IThreadingContext threadingContext,
+    IUIThreadOperationExecutor operationExecutor,
+    IAsynchronousOperationListener asyncListener,
+    Lazy<IStreamingFindUsagesPresenter> streamingPresenter) : INavigationActionFactory
+{
+    public Action CreateNavigationAction(string navigationTarget)
+    {
+        // ⚠ PERF: Avoid capturing Solution (including indirectly through Project or Document
+        // instances) as part of the navigationAction delegate.
+        var workspace = document.Project.Solution.Workspace;
+        var documentId = document.Id;
+
+        if (Uri.TryCreate(navigationTarget, UriKind.Absolute, out var absoluteUri))
+        {
+            var target = new QuickInfoHyperLink(workspace, absoluteUri);
+            return target.NavigationAction;
+        }
+
+        return () => NavigateToTargetAsync(
+            navigationTarget, workspace, documentId, threadingContext, operationExecutor, asyncListener, streamingPresenter.Value).Forget();
+    }
+
+    private static async Task NavigateToTargetAsync(
```

Comment:
📝 This method was moved from https://github.com/dotnet/roslyn/blob/4da55b7a6832f4b4f882c0f0a318ba42575fe0b1/src/EditorFeatures/Core/IntelliSense/Helpers.cs#L190-L232.